src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonICollectionOfTConverter.cs

```diff
@@ -0,0 +1,110 @@
+    /// </summary>
+    internal sealed class JsonICollectionOfTConverter<TCollection, TElement>
+        : JsonIEnumerableDefaultConverter<TCollection, TElement>
+        where TCollection : ICollection<TElement>
+    {
+        protected override void Add(TElement value, ref ReadStack state)
+        {
+            ((ICollection<TElement>)state.Current.ReturnValue!).Add(value);
+        }
+
+        protected override void CreateCollection(ref ReadStack state, JsonSerializerOptions options)
+        {
+            JsonClassInfo classInfo = state.Current.JsonClassInfo;
+
+            if ((TypeToConvert.IsInterface || TypeToConvert.IsAbstract))
+            {
+                if (!TypeToConvert.IsAssignableFrom(RuntimeType))
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(TypeToConvert);
+                }
+
+                state.Current.ReturnValue = new List<TElement>();
+            }
+            else
+            {
+                if (classInfo.CreateObject == null)
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(TypeToConvert);
+                }
+                else
+                {
+                    TCollection returnValue = (TCollection)classInfo.CreateObject!()!;
+
+                    if (returnValue.IsReadOnly)
+                    {
+                        ThrowHelper.ThrowNotSupportedException_SerializationNotSupportedCollection(TypeToConvert);
+                    }
+
+                    state.Current.ReturnValue = returnValue;
+                }
+            }
+        }
+
+        protected override bool OnWriteResume(
+            Utf8JsonWriter writer,
+            TCollection value,
+            JsonSerializerOptions options,
+            ref WriteStack state)
+        {
+            IEnumerator<TElement> enumerator;
+            if (state.Current.CollectionEnumerator == null)
+            {
+                enumerator = value.GetEnumerator();
+                if (!enumerator.MoveNext())
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                enumerator = (IEnumerator<TElement>)state.Current.CollectionEnumerator;
+            }
+
+            JsonConverter<TElement> converter = GetElementConverter(ref state);
+            do
+            {
+                if (ShouldFlush(writer, ref state))
+                {
+                    state.Current.CollectionEnumerator = enumerator;
+                    return false;
+                }
+
+                TElement element = enumerator.Current;
+                if (!converter.TryWrite(writer, element, options, ref state))
+                {
+                    state.Current.CollectionEnumerator = enumerator;
+                    return false;
+                }
+
+                state.Current.EndElement();
```

Comment:
For all collection converters, do we really need to call this? this sets `WriteStackFrame.PropertyState = None` which I don't see is being used on collections, only on objects and dictionaries.

Reply:
For non-dictionary collections, no currently we don't need it. I will remove it and rename `EndElement` to `EndDictionaryElement`. Thanks.

Btw we could also improve `EndProperty` to make a fast version for the "fast path" since it doesn't need to clear the state variables either. I won't do that yet until I can measure any noticable perf gain as there is a slight maintainability\understandability cost with that.